### PR TITLE
Deploy zipped assets to s3

### PR DESCRIPTION
**Pull request by piotrze**
Fixes: https://github.com/mdyd-dev/desksnearme/issues/8336

New option for `deploy` command
`marketplace-kit deploy staging --direct-assets-upload`

## QA
You can install cli from this PR branch with `npm install -g mdyd-dev/marketplace-kit#deploy-zipped-assets-to-s3`

- check `deploy --help` if it lists new option
- deploy some instance with assets
- deploy instance with big assets ex. 200MB 
- deploy instance with enormous amount of assets ex. 2GB
- deploy instance without new parameter

### CODE REVIEW

File: marketplace-kit-deploy.js [hunk 1]
```diff
@@ -7,45 +7,58 @@ const program = require('commander'),
...
+const deploy = async(env, authData, params) => {
+  await uploadArchive(env, params.directAssetsUpload);
+  if (params.directAssetsUpload){
+    const gateway = new Gateway(authData);
+    await assets.deployAssets(gateway);
+    logger.Success('Assets deployed to S3');
+  }
+};
+
+PARTNER_PORTAL_HOST = process.env.PARTNER_PORTAL_HOST || 'https://portal.apps.near-me.com';
```
Comment (pavelloz) [inline marketplace-kit-deploy.js]:
```suggestion
PARTNER_PORTAL_HOST = process.env.PARTNER_PORTAL_HOST || 'http://partners.platform-os.com';
```

Comment (piotrze) [reply marketplace-kit-deploy.js]: @pavelloz I've changed this address in separate PR.

File: lib/presignUrl.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  mime = require('mime');
+
+const presignUrl = (s3FileName, fileName) => {
+  const params = {
+    fileName: s3FileName,
+    contentLength: fs.statSync(fileName)['size'],
+    contentType: mime.getType(fileName)
+  };
+
+  const serviceUrl = `${process.env.DEPLOY_SERVICE_URL || process.env.MARKETPLACE_URL}/api/private/urls/presign-url`;
```
Comment (pavelloz) [inline lib/presignUrl.js]: Can be moved to line 5, its not dependent on anything in this func

File: lib/presignUrl.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      headers: {
+        token: process.env.MARKETPLACE_TOKEN,
+        marketplace_domain: url.parse(process.env.MARKETPLACE_URL).hostname,
+        marketplace_endpoint: PARTNER_PORTAL_HOST
+      },
+      qs: params,
+      json: true
+    }
+  ).then(body => {
+    return { uploadUrl: body.url, accessUrl: url.parse(body.accessUrl).href };
```
Comment (pavelloz) [inline lib/presignUrl.js]: // prettier-on

File: lib/assets.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    counter += 1;
+    if(fileExists) await sleep(1000);
+    fileExists = await request.head(fileUrl)
+      .then( () => true)
+      .catch({ statusCode: 403 }, () => false)
+      .catch((error) => logger.Error(error));
+  } while(fileExists && counter < 60);
+};
+
+const serializerManifestEntry = (file) => {
```
Comment (pavelloz) [inline lib/assets.js]: I would keep manifest specific methods in separate file.

File: lib/assets.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  glob = require('glob'),
+  request = require('request-promise'),
+  logger = require('./logger'),
+  prepareArchive = require('./prepareArchive'),
+  uploadFile = require('./s3UploadFile'),
+  presignUrl = require('./presignUrl'),
+  templates = require('./templates'),
+  settings = require('./settings');
+
+const addModulesToArchive = (archive) => {
```
Comment (pavelloz) [inline lib/assets.js]: Packing methods could also be extracted to cleanup

**Comment by pavelloz** [comment]
> marketplace-kit deploy staging --direct-assets-upload

`-d` ?

**Comment by jnorbert** [comment]
After deploying with a new command, there is the issue on the page

https://take.ms/scjx2

**Comment by jnorbert** [comment]
@piotrze I've noticed issue during deploy website with large size of assets

https://take.ms/K3VTe

And I am not sure why at the beginning is always the same information about the size of the archive - 0.54 MB

Another issue which I found is that no matter how big the archive is on the terminal is always information that deploys took 30 seconds -> Deploy succeeded after 0:30

**Pull request merged**